Download transfers starting from latest block header

dshulyak opened the pull request:
This change implements service that downloads transfers for specified ethereum address. Transfers include ethereum and erc20 transfers.

When user logs in we will start a reactor that will be periodically polling block chain for new blocks. Reactor will work transparently with any kind of node, it can be either local node or remote. Once we will receive header that wasn't previously seen we will check if there are any transfers attached to that block.

For ethereum we will compare balances between previous block and new one. If there is a difference we will download body of the block and iterate over transactions comparing recipient or signer of the transaction. If any will match such transaction will qualify as a transfer.

For erc20 we will use logs with erc20 Transfer events. Two separate queries will be sent for inbound and outbound events.

On pow chain reorgs are common thus we will have to check if new block parent matches to our previous known block. If they don't we will iterate backwards until we find common ancestor. All transfers that are attached to removed blocks will be also removed from database.

Transfers will be stored in sqlite database. It is not encrypted but path will be based on hex of the wallet address. 

API for client interaction described https://github.com/status-im/status-go/tree/71b6129b03b222e54880efe3132066a25e6c5e5e/services/wallet

dshulyak committed:
Download transfers starting from latest block header

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (30)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | e9c470f0 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/1/) | 2019-05-23 09:34:06 | ~51 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190523-093314-e9c470.zip) |
| :heavy_check_mark: | e9c470f0 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/1/) | 2019-05-23 09:36:53 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190523-093311-e9c470.zip) |
| :heavy_check_mark: | e9c470f0 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/1/) | 2019-05-23 09:44:12 | ~11 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190523-093310-e9c470.aar) |
| | | | | | | |
| :heavy_check_mark: | 73b81af7 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/2/) | 2019-05-24 08:42:19 | ~35 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190524-084144-73b81a.zip) |
| :heavy_check_mark: | 73b81af7 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/2/) | 2019-05-24 08:46:08 | ~4 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190524-084143-73b81a.zip) |
| :heavy_check_mark: | 73b81af7 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/2/) | 2019-05-24 08:48:03 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190524-084141-73b81a.aar) |
| | | | | | | |
| :heavy_check_mark: | ac760fc3 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/3/) | 2019-05-27 10:59:22 | ~46 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190527-105836-ac760f.zip) |
| :heavy_check_mark: | ac760fc3 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/3/) | 2019-05-27 11:02:46 | ~4 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190527-105835-ac760f.zip) |
| :heavy_check_mark: | ac760fc3 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/3/) | 2019-05-27 11:04:43 | ~6 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190527-105833-ac760f.aar) |
| | | | | | | |
| :heavy_check_mark: | def52f4b | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/4/) | 2019-05-27 11:26:43 | ~38 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190527-112604-def52f.zip) |
| :heavy_check_mark: | def52f4b | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/4/) | 2019-05-27 11:29:57 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190527-112602-def52f.zip) |
| :heavy_check_mark: | def52f4b | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/4/) | 2019-05-27 11:31:32 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190527-112600-def52f.aar) |
| | | | | | | |
| :heavy_check_mark: | 71b6129b | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/5/) | 2019-05-28 06:33:14 | ~36 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190528-063238-71b612.zip) |
| :heavy_check_mark: | 71b6129b | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/5/) | 2019-05-28 06:37:03 | ~4 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190528-063237-71b612.zip) |
| :heavy_check_mark: | 71b6129b | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/5/) | 2019-05-28 06:37:51 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190528-063235-71b612.aar) |
| | | | | | | |
| :heavy_check_mark: | d3cf6b34 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/6/) | 2019-05-31 06:21:46 | ~1 min | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190531-062031-d3cf6b.zip) |
| :heavy_check_mark: | d3cf6b34 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/6/) | 2019-05-31 06:23:43 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190531-062030-d3cf6b.zip) |
| :heavy_check_mark: | d3cf6b34 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/6/) | 2019-05-31 06:25:36 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190531-062029-d3cf6b.aar) |
| | | | | | | |
| :heavy_check_mark: | 275acd60 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/7/) | 2019-05-31 06:25:49 | ~56 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190531-062453-275acd.zip) |
| :heavy_check_mark: | 275acd60 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/7/) | 2019-05-31 06:27:45 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190531-062451-275acd.zip) |
| :heavy_check_mark: | 275acd60 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/7/) | 2019-05-31 06:30:46 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190531-062537-275acd.aar) |
| | | | | | | |
| :x: | 5c7dbf33 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/8/) | 2019-06-13 06:27:27 | ~9.8 sec | `android` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/8/consoleText) |
| :x: | 5c7dbf33 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/8/) | 2019-06-13 06:27:28 | ~8.8 sec | `linux` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/8/consoleText) |
| :x: | 5c7dbf33 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/8/) | 2019-06-13 06:27:29 | ~10 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/8/consoleText) |
| | | | | | | |
| :x: | 0b505c6e | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/9/) | 2019-06-13 06:29:04 | ~10 sec | `android` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/9/consoleText) |
| :x: | 0b505c6e | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/9/) | 2019-06-13 06:29:05 | ~9.9 sec | `linux` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/9/consoleText) |
| :x: | 0b505c6e | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/9/) | 2019-06-13 06:29:05 | ~10 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/9/consoleText) |
| | | | | | | |
| :x: | d5fca30a | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/10/) | 2019-06-14 04:53:20 | ~31 sec | `android` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/10/consoleText) |
| :x: | d5fca30a | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/10/) | 2019-06-14 04:53:20 | ~30 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/10/consoleText) |
| :heavy_check_mark: | d5fca30a | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/10/) | 2019-06-14 04:53:27 | ~36 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190614-045250-d5fca3.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 9d96bed9 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/11/) | 2019-06-14 08:39:22 | ~38 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190614-083844-9d96be.zip) |
| :heavy_check_mark: | 9d96bed9 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/11/) | 2019-06-14 08:42:26 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190614-083843-9d96be.zip) |
| :heavy_check_mark: | 9d96bed9 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/11/) | 2019-06-14 08:44:35 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190614-083841-9d96be.aar) |
| | | | | | | |
| :heavy_check_mark: | 37e0ad9b | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1467/12/) | 2019-06-14 09:00:20 | ~33 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190614-085947-37e0ad.zip) |
| :heavy_check_mark: | 37e0ad9b | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1467/12/) | 2019-06-14 09:02:16 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190614-085946-37e0ad.zip) |
| :heavy_check_mark: | 37e0ad9b | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1467/12/) | 2019-06-14 09:04:52 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190614-085944-37e0ad.aar) |

dshulyak committed:
More tests, service, api and signals

dshulyak committed:
Doc strings and cleanup

dshulyak committed:
Implement test for atomically adding and removing blocks

dshulyak committed:
Integrate wallet service with status node

dshulyak committed:
Add e2e tests for wallet service

dshulyak committed:
Add bindata.go to golanci exclusion list

dshulyak committed:
Provide readme wallet service

dshulyak committed:
Send 9 more transactions

dshulyak committed:
Use JSONBlob everywhere

dshulyak committed:
Add test with removing all added block due to reorg

dshulyak commented:
@yenda i am not sure if you got my messages in status. so, you can use ios/android builds from this branch for development. this version doesn't download historical transfers, but downloads new. 

described basic API in https://github.com/status-im/status-go/tree/71b6129b03b222e54880efe3132066a25e6c5e5e/services/wallet

adambabik commented on api/backend.go:
```diff
@@ -539,20 +555,45 @@ func (b *StatusBackend) SelectAccount(walletAddress, chatAddress, password strin
+}
 
+func (b *StatusBackend) startWallet() error {
+	if b.statusNode.Config().WalletConfig.Enabled {
```
Inverse? 
```suggestion
	if !b.statusNode.Config().WalletConfig.Enabled {
	  return nil
	}
```

adambabik commented on services/wallet/api.go:
```diff
@@ -0,0 +1,30 @@
+// return header, tx and receipt as a raw json.
+func (api *API) GetTransfers(ctx context.Context, start, end *big.Int) ([]Transfer, error) {
+	log.Debug("call to get transfers", "start", start, "end", end)
+	if api.s.db == nil {
```
That's ok but even better if we can hide the presence of `db` from the `API` completely. So, maybe `api.s.GetTransfers(start, end)` and then serialize it if needed?

dshulyak replied:
i actually don't want to add more methods to Service. initially i wanted to pass db to API and make sure that it doesn't have access to Service at all. but then i had to create db after API is created

adambabik replied:
Not sure if it matters here but generally it's better to have business logic in a service and API being just a wrapper. Then, you can use the service in Go and the API layer if going through JSON-RPC. If you put everything in the API, it might be inefficient for Go due to limitation of the API methods.

If you don't think that is the case here, let's keep it.

dshulyak replied:
i am not sure that i like when service implements all this logic. if i had to write multi-component service i would probably leave service to manage lifecycle and implement p2p interface. all other logic would be in sub-components, grouped by function for example Read and Send otr by objects that are used in each component. 

in general there is no problem using API directly in golang, one exception is when rpc.Subscription is returned. but if thats the issue there could be one more layer between service and api

dshulyak committed:
Implement downloaders for historical transfers

pedropombeiro commented on api/backend.go:
```diff
@@ -539,20 +555,45 @@ func (b *StatusBackend) SelectAccount(walletAddress, chatAddress, password strin
+			return err
+		}
+		path := path.Join(b.statusNode.Config().DataDir, fmt.Sprintf("wallet-%x.sql", account.Address))
+		err = wallet.StartReactor(path,
```
Why the use of `reactor` as a name instead of `monitor` (just seems a much more common name for this type of functionality)?

dshulyak replied:
i just found that Reactor is a common name for similar parts of applications across different projects and seems like a better name

pedropombeiro replied:
It's the first time I've seen it, it makes me think of a nuclear reactor :-)

pedropombeiro commented on services/wallet/README.md:
```diff
@@ -0,0 +1,121 @@
+==========
+
+Wallet service starts a loop that watches for new transfers (eth and erc20).
+To correctly start a service two values need to be changed in the config:
```
```suggestion
To correctly start the service, two values need to be changed in the config:
```

pedropombeiro commented on services/wallet/README.md:
````diff
@@ -0,0 +1,121 @@
+}
+```
+
+Wallet API
````
Again Wallet API section?

pedropombeiro commented on services/wallet/README.md:
```diff
@@ -0,0 +1,121 @@
+Signals
+-------
+
+Two signals will are emmited:
```
```suggestion
Two signals will are emitted:
```

pedropombeiro commented on services/wallet/README.md:
```diff
@@ -0,0 +1,121 @@
+
+Two signals will are emmited:
+
+1. Signal `newblock`
```
```suggestion
1. `newblock` signal
```

pedropombeiro commented on services/wallet/README.md:
```diff
@@ -0,0 +1,121 @@
+
+1. Signal `newblock`
+
+Emmited when transfers from new block were added to the database. In this case block number if the number of this new block.
```
```suggestion
Emitted when transfers from a new block were added to the database. In this case the block number is the number of this new block.
```

pedropombeiro commented on services/wallet/README.md:
````diff
@@ -0,0 +1,121 @@
+}
+```
+
+2. Signal `reorg`.
````
```suggestion
2. `reorg` signal
```

pedropombeiro commented on services/wallet/README.md:
```diff
@@ -0,0 +1,121 @@
+
+2. Signal `reorg`.
+
+Emmited when part of blocks were removed. Starting from a given block number all transfers were removed.
```
```suggestion
Emitted when part of blocks were removed. Starting from a given block number, all transfers were removed.
```

adambabik replied:
We should add information what is an expected action after each signal. Whether a call to `wallet_getTransfers` should happen and with what params from received signal.

dshulyak replied:
that was pretty much the whole idea. added some details to readme

pedropombeiro commented on services/wallet/database.go:
```diff
@@ -0,0 +1,254 @@
+	return err
+}
+
+// GetTransfers load transfers transfer betweeen two blocks.
```
```suggestion
// GetTransfers returns transfers that happened between two blocks.
```

pedropombeiro commented on services/wallet/downloader.go:
```diff
@@ -0,0 +1,159 @@
+	signer  types.Signer
+}
+
+// GetTransfers checks is balance was changed between two blocks. And if so it downloads transfers
```
```suggestion
// GetTransfers checks if the balance was changed between two blocks. if so, it downloads transfers
```

pedropombeiro commented on services/wallet/events.go:
```diff
@@ -0,0 +1,19 @@
+const (
+	// EventNewBlock emitted when new block was added to the same canonical chan.
+	EventNewBlock = "newblock"
+	// EventReorg emitted when canonical chain was changed. In this case BlockNumber will be an earliest added block.
```
```suggestion
	// EventReorg emitted when canonical chain was changed. In this case, BlockNumber will be a block added earlier.
```

dshulyak replied:
the point here is that there will be multiple blocks added in reorg, e.g. 4,5,6,7 and BlockNumber will be 4 - earliest added block. maybe there is a better word for it then earliest..

pedropombeiro replied:
`In this case, BlockNumber represents the first reorganized block.`?

dshulyak replied:
thanks, definitely better than what i wrote. i will change it in a followup PR to avoid rebase

pedropombeiro commented on services/wallet/service.go:
```diff
@@ -0,0 +1,84 @@
+	return s.signals.Start()
+}
+
+// StartReactor separately cause it requires known ethereum address, that will become available only after login.
```
```suggestion
// StartReactor separately because it requires known ethereum address, which will become available only after login.
```

pedropombeiro commented on services/wallet/service.go:
```diff
@@ -0,0 +1,84 @@
+	return s.db.Close()
+}
+
+// Stop reactor, signals  transmitter and close db.
```
```suggestion
// Stop reactor, signals transmitter and close db.
```

pedropombeiro commented on services/wallet/transmitter.go:
```diff
@@ -0,0 +1,64 @@
+	Subscribe(interface{}) event.Subscription
+}
+
+// SignalsTransmitter transmits received events as a wallet signals.
```
```suggestion
// SignalsTransmitter transmits received events as wallet signals.
```

pedropombeiro reviewed (commented):
wow, lots of work there! Added some minor comments.

adambabik commented on api/backend.go:
```diff
@@ -474,7 +475,19 @@ func (b *StatusBackend) Logout() error {
+		switch err {
+		case node.ErrServiceUnknown:
+		case nil:
+			err = wallet.StopReactor()
```
```suggestion
			return wallet.StopReactor()
```

dshulyak replied:
i can't because i need to exit only if err != nil. if it is nil then Logout needs to happen

adambabik commented on api/backend.go:
```diff
@@ -497,7 +510,10 @@ func (b *StatusBackend) reSelectAccount() error {
 		return err
 	}
-
+	err = b.startWallet()
```
```suggestion
	return b.startWallet()
```

adambabik commented on api/backend.go:
```diff
@@ -539,20 +555,45 @@ func (b *StatusBackend) SelectAccount(walletAddress, chatAddress, password strin
 			return err
 		}
 	}
+	err = b.startWallet()
```
```suggestion
	return b.startWallet()
```

adambabik commented on api/backend.go:
```diff
@@ -539,20 +555,45 @@ func (b *StatusBackend) SelectAccount(walletAddress, chatAddress, password strin
+			return err
+		}
+		path := path.Join(b.statusNode.Config().DataDir, fmt.Sprintf("wallet-%x.sql", account.Address))
+		err = wallet.StartReactor(path,
```
```suggestion
		return wallet.StartReactor(path,
```

adambabik commented on services/wallet/api.go:
```diff
@@ -0,0 +1,30 @@
+	if api.s.db == nil {
+		return nil, errors.New("wallet service is not initialized")
+	}
+	rst, err := api.s.db.GetTransfers(start, end)
```
```suggestion
	return api.s.db.GetTransfers(start, end)
```

dshulyak replied:
i prefer to return no transfers explicitly to avoid any confusion

adambabik commented on services/wallet/database.go:
```diff
@@ -0,0 +1,254 @@
+
+// JSONBlob type for marshaling/unmarshaling inner type to json.
+type JSONBlob struct {
+	Interface interface{}
```
```suggestion
	data interface{}
```

Because underlying value shouldn't be really exposed and `data` because it better describes the value.

dshulyak replied:
done

adambabik commented on services/wallet/database.go:
```diff
@@ -0,0 +1,254 @@
+	if err != nil {
+		return err
+	}
+	blocks, err = tx.Prepare("INSERT INTO blocks(hash,number,header) VALUES (?, ?, ?)")
```
```suggestion
	blocks, err = tx.Prepare("INSERT INTO blocks(hash, number, header) VALUES (?, ?, ?)")
```

adambabik commented on services/wallet/database.go:
```diff
@@ -0,0 +1,254 @@
+	return
+}
+
+// SaveHeader stores single header.
```
```suggestion
// SaveHeader stores a single header.
```

adambabik commented on services/wallet/database.go:
```diff
@@ -0,0 +1,254 @@
+	return err
+}
+
+// SaveHeaders atomically stores list of headers.
```
```suggestion
// SaveHeaders stores a list of headers atomically.
```

adambabik commented on services/wallet/database.go:
```diff
@@ -0,0 +1,254 @@
+	if err != nil {
+		return
+	}
+	defer func() {
```
It is repeated a few times:
```suggestion
	defer func() { err = closeTx(tx) }() 
```

dshulyak replied:
i tried. it looks much more readable the way it is now. note that i have to ignore error if tx is rollbacked and i need to pass error to closeTx to detemine how to close it

adambabik replied:
Ah right, it would look like `err = closeTx(tx, err)`. Not really clear.

adambabik commented on services/wallet/database.go:
```diff
@@ -0,0 +1,254 @@
+	if err == nil {
+		return header, nil
+	}
+	if err.Error() == errNoRows {
```
This special case is not tested.

dshulyak replied:
done

adambabik commented on services/wallet/downloader.go:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+	if len(rst) == 0 {
+		return nil, errors.New("balance changed but no new transactions were found")
```
Is this an error?

dshulyak replied:
yeah it is subtle but can happen. i query balance with block number and then query block body and because we are using infura we can hit different replica. it could be that we will hit replica A on first query and replica B on second. and block won't be the same there

but i realized that this is not the best way to test because gas spent will result in balance change as well

adambabik commented on services/wallet/erc20/erc20.go:
```diff
@@ -0,0 +1,353 @@
+// Code generated - DO NOT EDIT.
```
How is it generated? Do we have gogenerate for it?

dshulyak replied:
i described in README.md which solidity snippet was used to generate this file. i generated it with abigen. i can try to make gogenerate statement for it

dshulyak replied:
added file with gogenerate statement

adambabik commented on services/wallet/reactor.go:
```diff
@@ -0,0 +1,204 @@
+		// TODO(dshulyak) parametrize
+		// pow block on main chain is mined once per ~14 seconds
+		// but we for tests we are using clique chain with immediate block signer
+		ticker = time.NewTicker(1 * time.Second)
```
Let's solve it now with `Reactor` param. We can default it using a map of networkID to tick time.

dshulyak replied:
makes sense. will do it that way

adambabik reviewed (commented):
I would also add more comments to methods (even private ones) explaining why a given method exists and what it does with all cases described, even if the cases are not special. This should be understandable by people who don't know how Ethereum blockchain works.

dshulyak committed:
Store only hash and number of the block

dshulyak committed:
Implement iterative downloader

dshulyak committed:
Add tests that ensures that historical transfers will be included

dshulyak committed:
Send history event and update REDME

dshulyak committed:
Address review comments

dshulyak commented:
@PombeirP @adambabik thank you guys for review. pushed update with address comments

dshulyak committed:
Add more logs and adjust timeout after debuggin on main chain

dshulyak committed:
Implement faster downloader for historical eth transfers

dshulyak committed:
Use TransactionByHash instead of TransactionInBlock

dshulyak committed:
Adjust timeouts for erc20 downloader

dshulyak committed:
Transaction received in block can be nil

dshulyak committed:
Fix expectation in e2e test

dshulyak committed:
Check transaction.To after signer

dshulyak committed:
Raise waiting interval if there was an error in any loop

dshulyak committed:
Accept hex encoding in API and handle not specified start value

dshulyak committed:
Test binary iterative downloader

dshulyak committed:
Apply several suggestions after review

- More debug info when node is stopped
- low keys in events
- send history event only if new transfers were found

dshulyak committed:
Remove unused ethbatchSize

dshulyak committed:
Use parent ctx in iterators and hash of tx as a unique identifier

dshulyak committed:
Replace binary iterative downloader with concurrent downloader for eth

dshulyak committed:
Refactor logic from reactor

dshulyak committed:
Rework downloaders so that can accept multiple accounts

dshulyak committed:
Support for downloading transfers for multiple accounts

dshulyak committed:
Update readme with minor fixes

dshulyak committed:
Refactor issues reported by linter

dshulyak committed:
Rewrite concurrent downloader and make erc20 downloader concurrent

dshulyak committed:
Generate unique ID for every transfer

dshulyak committed:
Write better tests for reorg handling and improve handling description

dshulyak committed:
Renaming leftover

dshulyak committed:
Add logs and simplify lastKnownHeader

dshulyak committed:
Update accounts with all blocks in the set

dshulyak committed:
Encrypt database with same password taht is used for pfs db

dshulyak committed:
Rename ConcurrentRunner to AtomicGroup

dshulyak committed:
Add fast catchup after app was offline for longer than 15 blocks

When downloader is started it will go through following procedure:
- verify that the last synced block is still in canonical chain
- run fast indexer for every account from last synced block to head of the chain
- start watching block by block

dshulyak committed:
Remove ununes isSequence

dshulyak committed:
Increase timeout in test: bootstrap procedure may take longer now

dshulyak committed:
Remove unused code in database and adjust documentation

gravityblast commented:
@dshulyak would it be possible to expose a method to start the service accepting a list of addresses instead of using `b.accountManager.SelectedWalletAccount()`?

This would enable this part to work with multi-account, where `status-react` knows a list of addresses/accounts associated to a user.

dshulyak commented:
@gravityblast 

> @dshulyak would it be possible to expose a method to start the service accepting a list of addresses instead of using b.accountManager.SelectedWalletAccount()?

sure, can be separate method. maybe it makes sense to change signature of the SelectAccount and it will accept multiple wallet addresses? i will simply pass them to my service. but whatever you prefer

gravityblast commented:
With multi-account there will be watch-only addresses too, that are not even stored in go. So I think if we can have something callable from status-react it would be better. For keycard as well, addresses are not stored in go and keys are on the card.

dshulyak commented:
> With multi-account there will be watch-only addresses too, that are not even stored in go. So I think if we can have something callable from status-react it would be better. For keycard as well, addresses are not stored in go and keys are on the card.

thats clear. i meant that we can receive list of addresses to watch using the same method (SelectAccount), doesn't matter where they are stored. but if it makes more sense to have separate method than sure let's add it

gravityblast commented:
@dshulyak ok so status-react will login passing "main address"/password and a list of addresses to watch that will be set in a slice inside account manager. would that be ok?

dshulyak commented:
> @dshulyak ok so status-react will login passing "main address"/password and a list of addresses to watch that will be set in a slice inside account manager. would that be ok?

yes

gravityblast commented:
ok it makes sense! if we can already setting up to be compatible with multiple accounts then I'll update it when I change the login part, thank you!

dshulyak committed:
Use migrate/v4 instead of migrate

dshulyak committed:
Fix spelling
